### PR TITLE
[OIDC] Load configuration via jenkins.yaml

### PR DESCRIPTION
**Pull request by gaiksaya**
Signed-off-by: Sayali Gaikawad <gaiksaya@amazon.com>

### Description
Please see issue #85 for more details. We want to load the configuration via source coded config file so that the redeployment of the EC2 instance does not result in loss of config change.  This will also ensure that we make all changes to configuration only via code and not via UI.
The baseYaml remains the same and as per the features, a new file is created and loaded in jenkins.
I also think we should increment the version of this package as this is a breaking change.
Note: Until #88  is fixed, this change breaks agent node configuration.
 
### Issues Resolved
Part of https://github.com/opensearch-project/opensearch-ci/issues/88
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: lib/compute/jenkins-main-node.ts [hunk 1]
```diff
@@ -191,6 +198,8 @@ export class JenkinsMainNode {
       InitPackage.yum('java-1.8.0-openjdk-devel'),
       InitPackage.yum('openssl'),
       InitPackage.yum('mod_ssl'),
+      // eslint-disable-next-line max-len
+      InitCommand.shellCommand('sudo wget -nv https://github.com/mikefarah/yq/releases/download/v4.22.1/yq_linux_amd64 -O /usr/bin/yq && sudo chmod +x /usr/bin/yq'),
```
Comment (abhinavGupta16) [inline lib/compute/jenkins-main-node.ts:202]: That is this repository used for?

File: lib/compute/jenkins-main-node.ts [hunk 2]
```diff
@@ -368,71 +377,31 @@ export class JenkinsMainNode {
...
-        + ' xmlstarlet ed -L -d "/hudson/securityRealm"'
-        + ' -s /hudson -t elem -n securityRealm -v " "'
-        + ' -i //securityRealm -t attr -n "class" -v "org.jenkinsci.plugins.oic.OicSecurityRealm"'
-        + ' -i //securityRealm -t attr -n "plugin" -v "oic-auth@1.8"'
+        + ' varkeys=`echo $var | yq \'keys\' | cut -d "-" -f2 | cut -d " " -f2` &&'
         // eslint-disable-next-line max-len
-        + `${JenkinsMainNodeConfig.oidcConfigFields().map((e) => ` -s /hudson/securityRealm -t elem -n ${e[0]} -v ${e[1] === 'replace' ? `"$(echo $var | jq -r ".${e[0]}")"` : `"${e[1]}"`}`).join(' ')}`
-        + ' /var/lib/jenkins/config.xml'
-        + ` && ${RELOAD_CONFIG}`
+        + ' for i in $varkeys; do newvalue=`echo $var | yq .$i` && myenv=$newvalue i=$i yq -i \'.jenkins.securityRealm.oic.[env(i)]=env(myenv)\' /var/lib/jenkins/jenkins.yaml ; done'
```
Comment (abhinavGupta16) [inline lib/compute/jenkins-main-node.ts:387]: Shouldn't we parse the yml file instead? How is this different from xmlstarlet?

File: .gitignore [hunk 3]
```diff
@@ -3,6 +3,8 @@
 !jest.config.js
 *.d.ts
 node_modules
+resources/jenkins.yaml
```
Comment (abhinavGupta16) [inline .gitignore:6]: Can you add a comment here explaining why are we ignoring this file?

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]

File: lib/compute/jenkins-main-node.ts (hunk 1 above)
Comment (gaiksaya) [reply lib/compute/jenkins-main-node.ts:202]: You cannot add aws secrets values in typescript. I am getting values and replacing them in already loaded yaml on the server

File: lib/compute/jenkins-main-node.ts (hunk 2 above)
Comment (gaiksaya) [reply lib/compute/jenkins-main-node.ts:387]: Please refer the comment above. we are not adding any config via this command, Just replacing with user provided values

File: .gitignore (hunk 3 above)
Comment (gaiksaya) [reply .gitignore:6]: will do

File: lib/compute/jenkins-main-node.ts (hunk 2 above)
Comment (abhinavGupta16) [reply lib/compute/jenkins-main-node.ts:387]: This is exactly what we can do with xml as well right?

File: test/data/jenkins.yaml [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    tokenGenerationOnCreationEnabled: false
+    usageStatisticsEnabled: true
+  copyartifact:
+    mode: PRODUCTION
+  globalJobDslSecurityConfiguration:
+    useScriptSecurity: true
+  sSHD:
+    port: -1
+unclassified:
+  ansiColorBuildWrapper:
```
Comment (prudhvigodithi) [inline test/data/jenkins.yaml:84]: we can remove this ansiColorBuildWrapper as by default this be coming default from ANSI plugin.

File: lib/compute/jenkins-main-node.ts (hunk 1 above)
Comment (abhinavGupta16) [reply lib/compute/jenkins-main-node.ts:202]: Can we rather extract the values in EC2, return them and use them in code?

File: .gitignore (hunk 3 above)
Comment (abhinavGupta16) [reply .gitignore:6]: thanks

File: lib/compute/jenkins-main-node.ts (hunk 1 above)
Comment (gaiksaya) [reply lib/compute/jenkins-main-node.ts:202]: Nope, that will over complicate this unnecessarily. Plus from security point of view it is better that the secrets stay on server than written to local file.

File: lib/compute/jenkins-main-node.ts (hunk 2 above)
Comment (gaiksaya) [reply lib/compute/jenkins-main-node.ts:387]: We are adding configuration via code (yaml in this case). You can check linked issues for details why we reached this decision. Also editing, maintaining and manipulating yaml is much easier than xml. Now on, every user does not have to learn xmlstartlet to add code to xml. It was a very messy process and changes in hudson libraries and config might have affected us. Right now, yaml you add the config you want in yaml and yaml takes care of it to add it to config. xml.

TL;DR:
Easy to maintain, codeify and update the config via code. (It was like we were editing bytecodes by editing config.xml which will have serious implications and unknown errors.)

File: .gitignore (hunk 3 above)
Comment (gaiksaya) [reply .gitignore:7]: Added

File: test/data/jenkins.yaml (hunk 4 above)
Comment (gaiksaya) [reply test/data/jenkins.yaml:84]: Removed.

File: lib/compute/jenkins-main-node.ts (hunk 2 above)
Comment (abhinavGupta16) [reply lib/compute/jenkins-main-node.ts:387]: Do we have to learn `yq` now?

File: lib/compute/jenkins-main-node.ts (hunk 1 above)
Comment (abhinavGupta16) [reply lib/compute/jenkins-main-node.ts:202]: The secret would still be written on the file on the server, just like now.

**Review by abhinavGupta16 — APPROVED** [review]

File: lib/compute/jenkins-main-node.ts (hunk 2 above)
Comment (gaiksaya) [reply lib/compute/jenkins-main-node.ts:387]: Just for replacing any values in jenkins.yaml on the jenkins server directly during installation.

File: lib/compute/jenkins-main-node.ts (hunk 1 above)
Comment (abhinavGupta16) [reply lib/compute/jenkins-main-node.ts:202]: discussed offline
